neuromaps/nulls/nulls.py

```diff
@@ -409,56 +428,99 @@ def _surf_surrogates(data, atlas, density, parcellation, distmat, n_proc):
+    if parcellation is None:
+        if nib.load(atlas['2009cAsym_T1w']).shape == darr.shape:
+            darr *= load_data(atlas['2009cAsym_brainmask'])
+        elif ((density == '1mm' or density == '2mm')
+                and nib.load(atlas['6Asym_T1w']).shape == darr.shape):
+            darr *= load_data(atlas['6Asym_brainmask'])
+        else:
+            darr *= load_data(
+                mni152_to_mni152(nib.load(atlas['2009cAsym_brainmask']),
+                                 data,
+                                 'nearest')
+            )
+        mask = np.logical_not(
+            np.logical_or(np.isclose(darr, 0), np.isnan(darr))
+        )
+        xyz = nib.affines.apply_affine(load_nifti(data).affine,
+                                       np.column_stack(np.where(mask)))
+    else:
+        parr = load_data(parcellation)
+        labels = np.trim_zeros(np.unique(parr))
+        mask = np.logical_not(
+            np.logical_or(np.isclose(parr, 0), np.isnan(parr))
+        )
+        xyz = nib.affines.apply_affine(load_nifti(parcellation).affine,
+                                       np.column_stack(np.where(mask)))
+        parcellation = parr[mask]
 
     # calculate distance matrix
     index = None
     if distmat is None:
         if parcellation is None:  # memmap because BIG
             distout, indout = tmpname('.mmap'), tmpname('.mmap')
-            dist = np.lib.format.open_memmap(distout, mode='w+',
-                                             dtype='float32',
-                                             shape=(len(xyz), len(xyz)))
-            index = np.lib.format.open_memmap(indout, mode='w+',
-                                              dtype='int32',
-                                              shape=(len(xyz), len(xyz)))
+            try:
+                dist = np.lib.format.open_memmap(distout, mode='w+',
+                                                 dtype='float32',
+                                                 shape=(len(xyz), len(xyz)))
+                index = np.lib.format.open_memmap(indout, mode='w+',
+                                                  dtype='int32',
+                                                  shape=(len(xyz), len(xyz)))
+                for n, row in enumerate(xyz):
+                    xyz_dist = cdist(row[None], xyz).astype('float32')
+                    index[n] = np.argsort(xyz_dist, axis=-1).astype('int32')
+                    dist[n] = xyz_dist[:, index[n]]
+                    dist.flush()
+                    index.flush()
+            except Exception as err:
+                if 'dist' in locals():
+                    dist._mmap.close()
+                    dist.filename.unlink()
+                    del dist
+                if 'index' in locals() and index is not None:
+                    index._mmap.close()
+                    index.filename.unlink()
+                    del index
+                raise err
```

Comment:
I know that this `try/except` clause is not ideal, but the main issue with the code is that currently, when an error is raised, the mmap files aren't deleted. It can be very problematic if the user isn't aware that those files got created in their` $TEMP$` folder. They are usually more than 20Gb in size so I think that we should be really careful when handling them. 

My rationale at the time was that since we don't really want to handle the raised expection, and since, to answer your second question, I don't know what type of error could be raised in there, we can allow ourselves to be non-specific, especially since the `err` message is raised inside the except block.  

And, actually, according to my rationale, we should perhaps even do `except BaseException`, which is even less recommended :sweat_smile: since currently, the except block does not catch `KeyboardInterrupt` exceptions.

Reply:
Hmmm, okay. I still don't love it, but I get where you're coming from—those mmap files are a huge pain in the butt.

At the very least, when the mmap files are first _created_ they are empty and therefore take up no actual disk space (your OS may lie to you and say it's actually using many many GBs of space—do not believe the computer, it is actually [mostly empty](https://stackoverflow.com/questions/34023665/numpy-memmap-bogus-memory-allocation/34029352#34029352)). As such, perhaps the creation of the mmap files can be moved out of the `try` block, in which case you can also remove the check for `if 'dist' in locals()` and `if 'index' in locals()` in the `except` block to make this all a little more palatable.

Reply:
Why don't we just use `tempfile.NamedTemporaryFile()` instead of `tempfile.mkstemp()` since the files are automatically deleted on L557 and L558 anyway? This way, the files would be automatically deleted, and we wouldn't need the `try/except` block. We could then also get rid of the `if hasattr(hdist, 'filename'): ...` block in the `_make_surrogates()` function.

Or alternatively, we could add a `save_dist` parameter in the `burt2020()` function and `burt2018()` and `moran()`), which would give the user the option of either saving the `dist` matrix or not. If they want to save it, then we don't need `try/except` blocks since we don't need to delete the file. And we'd use either `tempfile.NamedTemporaryFile()` or `tempfile.mkstemp()` depending on what the user chooses.

Reply:
Sure ! That works. I may have had a reason for doing it this way at some point but I think the `tempfile.NamedTemporaryFile` method seems better, so let's go for it. It'll definitely help clean up a lot of code.

Re `save_dist` parameter: yes, I am all here for it. I think you can still use `tempfile.NamedTemporaryFile` and then, if `save_dist` is a valid filepath or something (i.e., `save_dist is not None`), just pass `delete=False` when creating the file with `NamedTemporaryFile` and move it to wherever was specified !

Reply:
Yeah, I think that `NamedTemporaryFile` would be more suitable for our needs. However, an issue would be that we'd still need to close the file object since [relying on garbage collection is really not a good practice](https://stackoverflow.com/questions/1832528/is-close-necessary-when-using-iterator-on-a-python-file-object) (this was an issue when we were using `mkstemp()` as well).

So we'd either need a `try/finally` block or a `with` statement, but we can't really do that since we want to return the file object at the end of `_vol_surrogates()`, and not close it. 

There must be ways of doing it, but the only ones I have in mind right now would require us to refactor `_makes_surrogates()` as well. So I would be ok with the file object being closed (and therefore the memmaps being deleted) at garbage collection when the code is interupted.  

I'd however keep this block of code in `_make_surrogates()` so that the file object closes properly when the code is not interupted:

```
            if hasattr(hdist, 'filename'):
                hdist._mmap.close()
                hind._mmap.close()
                hdist.filename.unlink()
                hind.filename.unlink()
                del hdist, hind
```